components/raftstore/src/store/fsm/peer.rs

```diff
@@ -1481,6 +1515,23 @@ where
 
     #[inline]
     fn check_force_leader(&mut self) {
+        if let Some(ForceLeaderState::WaitTicks {
+            failed_stores,
+            ticks,
+        }) = &mut self.fsm.peer.force_leader
+        {
+            *ticks -= 1;
+            debug!(
+                "sub tick";
+                "ticks" => *ticks,
+            );
+            if *ticks == 0 {
+                let s = mem::take(failed_stores);
+                self.on_enter_pre_force_leader(s);
```

Comment:
Instead of calling back on `on_enter_pre_force_leader()` maybe it looks cleaner if we separate the `on_enter_pre_force_leader()` into several functions like

```
SignificantMsg::EnterForceLeaderState => {
    self.on_enter_force_leader();
}

fn on_enter_force_leader() {
    if state.is_some() {
        ...
    }
    if is_leader || （promotable && has_leader）{
        self.leader_absence_check()
    } else {
        self.voter_liveness_check()
    }
}

fn check_force_leader() {
    if leader_absence_check_passes {
        self.voter_liveness_check()
    }
    if voter_liveness_check_passes {
        ...
    }
}
```

Reply:
After waiting ticks, we need to re-check all the conditions in `on_enter_pre_force_leader`. Reusing is straightforward.

Reply:
That's also something we can optimize here, instead of waiting for certain ticks, and then check the leader absence in the  `on_enter_pre_force_leader()`, maybe we can proactively check whether the leader is absent or not in each tick at `check_force_leader()`?

Reply:
Can't. For leader, `CheckQuorum` needs to wait for a round to see if peers are recent active; for follower, only after an election timeout, it can make sure the leader lease is expired. Anyway, both of them need to wait for some ticks to get the correct information.